File: packages/logger/src/logger.ts
/**
 * Orchestra Logger
 * 符合永乐大典地基规范的日志器
 */

import pino, { Logger as PinoLogger } from 'pino';
import { LogLevel, LogContext, LoggerConfig, LOG_LEVEL_PRIORITY } from './types';

// 默认配置
const DEFAULT_CONFIG: LoggerConfig = {
  level: 'info',
  format: 'json',
  console: true,
  pretty: process.env.NODE_ENV === 'development',
  timestamp: true,
};

/**
 * Orchestra Logger 类
 */
export class OrchestraLogger {
  private logger: PinoLogger;
  private module: string;
  private defaultContext: LogContext;

  constructor(module: string, config?: LoggerConfig) {
    this.module = module;
    this.defaultContext = { module };

    const mergedConfig = { ...DEFAULT_CONFIG, ...config };

    // 创建 pino logger
    this.logger = pino({
      level: mergedConfig.level,
      formatters: {
        level: (label) => ({ level: label.toUpperCase() }),
      },
      timestamp: mergedConfig.timestamp
        ? () => `,"timestamp":"${new Date().toISOString()}"`
        : false,
      transport: mergedConfig.pretty
        ? {
            target: 'pino-pretty',
            options: {
              colorize: true,
              translateTime: 'SYS:standard',
              ignore: 'pid,hostname',
            },
          }
        : undefined,
    });
  }

  /**
   * 创建子 logger，继承上下文
   */
  child(context: LogContext): OrchestraLogger {
    const childLogger = Object.create(this) as OrchestraLogger;
    childLogger.defaultContext = { ...this.defaultContext, ...context };
    childLogger.logger = this.logger.child(context);
    return childLogger;
  }

  /**
   * 绑定默认上下文
   */
  bind(context: LogContext): OrchestraLogger {
    return this.child(context);
  }

  /**
   * Debug 级别日志
   */
  debug(message: string, context?: LogContext): void {
    this.log('debug', message, context);
  }

  /**
   * Info 级别日志
   */
  info(message: string, context?: LogContext): void {
    this.log('info', message, context);
  }

  /**
   * Warn 级别日志
   */
  warn(message: string, context?: LogContext): void {
    this.log('warn', message, context);
  }

  /**
   * Error 级别日志
   */
  error(message: string, context?: LogContext & { error?: Error }): void {
    const { error, ...rest } = context || {};
    const logContext: LogContext = {
      ...rest,
      ...(error && {
        error_name: error.name,
        error_message: error.message,
        error_stack: error.stack,
      }),
    };
    this.log('error', message, logContext);
  }

  /**
   * Fatal 级别日志
   */
  fatal(message: string, context?: LogContext & { error?: Error }): void {
    const { error, ...rest } = context || {};
    const logContext: LogContext = {
      ...rest,
      ...(error && {
        error_name: error.name,
        error_message: error.message,
        error_stack: error.stack,
      }),
    };
    this.log('fatal', message, logContext);
  }

  /**
   * 记录 API 请求
   */
  request(
    method: string,
    path: string,
    context?: LogContext & { statusCode?: number; durationMs?: number }
  ): void {
    const { statusCode, durationMs, ...rest } = context || {};
    this.info(`${method} ${path}`, {
      ...rest,
      http_method: method,
      http_path: path,
      http_status: statusCode,
      duration_ms: durationMs,
    });
  }

  /**
   * 记录数据库操作
   */
  db(
    operation: string,
    table: string,
    context?: LogContext & { durationMs?: number; rowCount?: number }
  ): void {
    const { durationMs, rowCount, ...rest } = context || {};
    this.debug(`DB ${operation} ${table}`, {
      ...rest,
      db_operation: operation,
      db_table: table,
      duration_ms: durationMs,
      row_count: rowCount,
    });
  }

  /**
   * 记录外部服务调用
   */
  external(
    service: string,
    operation: string,
    context?: LogContext & { durationMs?: number; success?: boolean }
  ): void {
    const { durationMs, success, ...rest } = context || {};
    const level = success === false ? 'warn' : 'info';
    this.log(level, `External: ${service}.${operation}`, {
      ...rest,
      external_service: service,
      external_operation: operation,
      duration_ms: durationMs,
      success,
    });
  }

  /**
   * 计时器 - 记录操作耗时
   */
  time(label: string): () => void {
    const start = Date.now();
    return () => {
      const durationMs = Date.now() - start;
      this.debug(`${label} completed`, { durationMs, label });
    };
  }

  /**
   * 异步计时器
   */
  async timeAsync<T>(label: string, fn: () => Promise<T>): Promise<T> {
    const start = Date.now();
    try {
      const result = await fn();
      const durationMs = Date.now() - start;
      this.debug(`${label} completed`, { durationMs, label, success: true });
      return result;
    } catch (error) {
      const durationMs = Date.now() - start;
      this.error(`${label} failed`, {
        durationMs,
        label,
        success: false,
        error: error as Error,
      });
      throw error;
    }
  }

  /**
   * 核心日志方法
   */
  private log(level: LogLevel, message: string, context?: LogContext): void {
    const mergedContext = {
      ...this.defaultContext,
      ...context,
    };

    // 提取特殊字段
    const { module, requestId, userId, ...rest } = mergedContext;

    const logObject = {
      module,
      ...(requestId && { request_id: requestId }),
      ...(userId && { user_id: userId }),
      ...rest,
    };

    this.logger[level](logObject, message);
  }
}

// ============ 工厂函数 ============

/**
 * 创建 logger 实例
 *
 * @example
 * const logger = createLogger('UserService');
 * logger.info('用户登录成功', { userId: '123' });
 */
export function createLogger(module: string, config?: LoggerConfig): OrchestraLogger {
  return new OrchestraLogger(module, config);
}

// ============ 全局 logger ============

let globalLogger: OrchestraLogger | null = null;

/**
 * 获取全局 logger
 */
export function getLogger(): OrchestraLogger {
  if (!globalLogger) {
    globalLogger = createLogger('App');
  }
  return globalLogger;
}

/**
 * 设置全局 logger
 */
export function setGlobalLogger(logger: OrchestraLogger): void {
  globalLogger = logger;
}

/**
 * 初始化全局 logger
 */
export function initLogger(module: string, config?: LoggerConfig): OrchestraLogger {
  globalLogger = createLogger(module, config);
  return globalLogger;
}

File: packages/python/orchestra_foundation/logger.py
"""
日志模块
符合永乐大典地基规范的日志器
"""

import sys
from datetime import datetime, timezone
from typing import Any, Optional, Dict
from loguru import logger as _loguru_logger

from .request_id import get_request_id, get_user_id


# ============ 配置 Loguru ============


def _setup_loguru(level: str = "INFO", format_type: str = "json") -> None:
    """配置 loguru"""
    _loguru_logger.remove()

    if format_type == "json":
        # JSON 格式
        format_str = (
            '{{"level":"{level}","timestamp":"{time:YYYY-MM-DDTHH:mm:ss.SSS}Z",'
            '"module":"{extra[module]}","message":"{message}"{extra[_context_str]}}}'
        )
    else:
        # 文本格式
        format_str = (
            "<green>{time:YYYY-MM-DD HH:mm:ss}</green> | "
            "<level>{level: <8}</level> | "
            "<cyan>{extra[module]}</cyan> | "
            "{message}"
        )

    _loguru_logger.add(
        sys.stdout,
        format=format_str,
        level=level,
        colorize=(format_type != "json"),
    )


# 默认配置
_setup_loguru()


# ============ Logger 类 ============


class OrchestraLogger:
    """永乐大典标准日志器"""

    def __init__(self, module: str, level: str = "INFO"):
        self.module = module
        self._logger = _loguru_logger.bind(module=module, _context_str="")

    def _format_context(self, context: Optional[Dict[str, Any]]) -> str:
        """格式化上下文为 JSON 字符串片段"""
        if not context:
            return ""

        # 自动添加 request_id 和 user_id
        ctx = dict(context)
        if "request_id" not in ctx:
            req_id = get_request_id()
            if req_id:
                ctx["request_id"] = req_id
        if "user_id" not in ctx:
            uid = get_user_id()
            if uid:
                ctx["user_id"] = uid

        if not ctx:
            return ""

        pairs = []
        for k, v in ctx.items():
            if isinstance(v, str):
                pairs.append(f'"{k}":"{v}"')
            elif isinstance(v, bool):
                pairs.append(f'"{k}":{str(v).lower()}')
            elif v is None:
                pairs.append(f'"{k}":null')
            else:
                pairs.append(f'"{k}":{v}')

        return "," + ",".join(pairs)

    def debug(self, message: str, **context: Any) -> None:
        """Debug 级别日志"""
        self._log("DEBUG", message, context)

    def info(self, message: str, **context: Any) -> None:
        """Info 级别日志"""
        self._log("INFO", message, context)

    def warn(self, message: str, **context: Any) -> None:
        """Warn 级别日志"""
        self._log("WARNING", message, context)

    def warning(self, message: str, **context: Any) -> None:
        """Warning 级别日志（alias）"""
        self.warn(message, **context)

    def error(self, message: str, error: Optional[Exception] = None, **context: Any) -> None:
        """Error 级别日志"""
        if error:
            context["error_type"] = type(error).__name__
            context["error_message"] = str(error)
        self._log("ERROR", message, context)

    def fatal(self, message: str, error: Optional[Exception] = None, **context: Any) -> None:
        """Fatal 级别日志"""
        if error:
            context["error_type"] = type(error).__name__
            context["error_message"] = str(error)
        self._log("CRITICAL", message, context)

    def _log(self, level: str, message: str, context: Dict[str, Any]) -> None:
        """核心日志方法"""
        context_str = self._format_context(context)
        log_method = getattr(self._logger.bind(_context_str=context_str), level.lower())
        log_method(message)

    # ============ 便捷方法 ============

    def request(
        self,
        method: str,
        path: str,
        status_code: Optional[int] = None,
        duration_ms: Optional[float] = None,
        **context: Any,
    ) -> None:
        """记录 HTTP 请求"""
        ctx = {
            "http_method": method,
            "http_path": path,
            **context,
        }
        if status_code is not None:
            ctx["http_status"] = status_code
        if duration_ms is not None:
            ctx["duration_ms"] = round(duration_ms, 2)

        self.info(f"{method} {path}", **ctx)

    def db(
        self,
        operation: str,
        table: str,
        duration_ms: Optional[float] = None,
        row_count: Optional[int] = None,
        **context: Any,
    ) -> None:
        """记录数据库操作"""
        ctx = {
            "db_operation": operation,
            "db_table": table,
            **context,
        }
        if duration_ms is not None:
            ctx["duration_ms"] = round(duration_ms, 2)
        if row_count is not None:
            ctx["row_count"] = row_count

        self.debug(f"DB {operation} {table}", **ctx)

    def external(
        self,
        service: str,
        operation: str,
        duration_ms: Optional[float] = None,
        success: bool = True,
        **context: Any,
    ) -> None:
        """记录外部服务调用"""
        ctx = {
            "external_service": service,
            "external_operation": operation,
            "success": success,
            **context,
        }
        if duration_ms is not None:
            ctx["duration_ms"] = round(duration_ms, 2)

        if success:
            self.info(f"External: {service}.{operation}", **ctx)
        else:
            self.warn(f"External: {service}.{operation}", **ctx)


# ============ 工厂函数 ============


def create_logger(module: str, level: str = "INFO") -> OrchestraLogger:
    """
    创建 logger 实例

    Example:
        >>> logger = create_logger("UserService")
        >>> logger.info("用户登录成功", user_id="123")
    """
    return OrchestraLogger(module, level)


# ============ 全局 logger ============

_global_logger: Optional[OrchestraLogger] = None


def get_logger() -> OrchestraLogger:
    """获取全局 logger"""
    global _global_logger
    if _global_logger is None:
        _global_logger = create_logger("App")
    return _global_logger


def init_logger(
    module: str = "App",
    level: str = "INFO",
    format_type: str = "json",
) -> OrchestraLogger:
    """
    初始化全局 logger

    Args:
        module: 模块名
        level: 日志级别 (DEBUG, INFO, WARNING, ERROR, CRITICAL)
        format_type: 输出格式 (json, text)
    """
    global _global_logger
    _setup_loguru(level, format_type)
    _global_logger = create_logger(module, level)
    return _global_logger
